homeassistant/components/elmo_alarm/__init__.py

```diff
@@ -0,0 +1,219 @@
+        self._username = username
+        self._password = password
+        self._states_config = states_config
+        self._data = None
+        self._states = None
+        self._state = None
+
+        ElmoClient.__init__(self, host, vendor)
+
+    async def update(self):
+        try:
+            data = self.check()
+        except Exception:
+            self.auth(self._username, self._password)
+            data = self.check()
+        if self._data is None:
+            await self._configure_states()
+        self._data = data
+        await self._update_arm_state()
+        await self._update_zone_state()
+        await self._update_input_state()
+
+    async def _configure_states(self):
+        state_config = {}
+        for state in self._states_config:
+            state_config[state[STATE_NAME]] = state[STATE_ZONES]
+        self._states = state_config
+
+    async def _update_arm_state(self):
+        areas_armed = self._data["areas_armed"]
+
+        if len(areas_armed) == 0:
+            self._state = STATE_ALARM_DISARMED
+        else:
+            armed_indexes = [x["index"] + 1 for x in areas_armed]
+
+            state = [
+                state
+                for state, areas in self._states.items()
+                if set(areas) == set(armed_indexes)
+            ]
+
+            if len(state) > 0:
+                state = state[0]
+                if state == "arm_away":
+                    self._state = STATE_ALARM_ARMED_AWAY
+                elif state == "arm_home":
+                    self._state = STATE_ALARM_ARMED_HOME
+                elif state == "arm_night":
+                    self._state = STATE_ALARM_ARMED_NIGHT
+                elif state == "arm_custom_bypass":
+                    self._state = STATE_ALARM_ARMED_CUSTOM_BYPASS
+                else:
+                    self._state = None
+            else:
+                self._state = None
+
+    async def _update_zone_state(self):
+        self._zones = [
+            ZoneData(zone_id=area["index"], zone_name=area["name"], state=ZONE_ARMED)
+            for area in self._data["areas_armed"]
+            if area["name"] != "Unknown"
+        ]
+        self._zones += [
+            ZoneData(zone_id=area["index"], zone_name=area["name"], state=ZONE_DISARMED)
+            for area in self._data["areas_disarmed"]
+            if area["name"] != "Unknown"
+        ]
+
+    async def _update_input_state(self):
+        self._inputs = [
+            InputData(input_id=inp["index"], input_name=inp["name"], state=INPUT_ALERT)
+            for inp in self._data["inputs_alerted"]
+            if inp["name"] != "Unknown"
+        ]
+        self._inputs += [
+            InputData(input_id=inp["index"], input_name=inp["name"], state=INPUT_WAIT)
+            for inp in self._data["inputs_wait"]
+            if inp["name"] != "Unknown"
+        ]
+
+
+async def async_setup(hass, config):
```

Comment:
Please move this above `class ElmoClientWrapper` to make it more readable.

Reply:
Done.